src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -518,20 +523,55 @@ public CheckConstraintsArgs(CSharpCompilation currentCompilation, ConversionsBas
             }
         }
 
+        private static readonly ObjectPool<CheckConstraintsArgsBoxed> s_checkConstraintsArgsBoxedPool = new ObjectPool<CheckConstraintsArgsBoxed>(() => new CheckConstraintsArgsBoxed());
+
         internal sealed class CheckConstraintsArgsBoxed
         {
-            public readonly CheckConstraintsArgs Args;
+            public CheckConstraintsArgs Args;
 
             [MethodImpl(MethodImplOptions.NoInlining)]
-            public CheckConstraintsArgsBoxed(CSharpCompilation currentCompilation, ConversionsBase conversions, Location location, BindingDiagnosticBag diagnostics)
+            public static PooledCheckConstraintsArgsBoxed Create(CSharpCompilation currentCompilation, ConversionsBase conversions, Location location, BindingDiagnosticBag diagnostics)
             {
-                Args = new CheckConstraintsArgs(currentCompilation, conversions, location, diagnostics);
+                var boxedArgs = s_checkConstraintsArgsBoxedPool.Allocate();
+                boxedArgs.Args = new CheckConstraintsArgs(currentCompilation, conversions, location, diagnostics);
+                return new PooledCheckConstraintsArgsBoxed(boxedArgs);
             }
 
             [MethodImpl(MethodImplOptions.NoInlining)]
-            public CheckConstraintsArgsBoxed(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics)
+            public static PooledCheckConstraintsArgsBoxed Create(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics)
```

Comment:
>PooledCheckConstraintsArgsBoxed Create [](http://example.com/codeflow?start=26&length=38)

It is weird (ie. unexpected and not very readable) to have a `Create` on one type that returns a different type.

Reply:
updated.